Tools/Scripts/hooks/pre-push

```diff
@@ -0,0 +1,323 @@
+        secondary = match.group('secondary')
+        if secondary:
+            secondary = UNPACK_SECONDARY_RE.match(secondary).groups()[0]
+        for candidate in [primary, secondary]:
+            original_remotes = remotes_for(candidate)
+            if not original_remotes:
+                continue
+            return security_level_for_remotes(original_remotes, name_to_remote), CLASS_2
+        break
+
+    # Class 3: We must inspect the commit for bug references
+    if REPOSITORY:
+        obj = Commit(hash=commit, message=message)
+        for issue in obj.issues:
+            if issue.redacted:
+                return 1, CLASS_3
+        return 0, CLASS_3
+
+    return None, None
+
+
+def main(name, remote):
+    target_level = security_level_for_remote(name, remote)
+    if VERBOSITY > QUIET:
+        if target_level:
+            print('Pushing to {}, which is classified as a secure remote'.format(name or remote))
+        elif target_level == 0:
+            print('Pushing to {}, which is classified as a public remote'.format(name or remote))
+        else:
+            print('Pushing to {}, which has not classification. Treating it as a public remote'.format(name or remote))
+
+    to_push = []
+    for push in sys.stdin.readlines():
+        _, local_ref, _, remote_ref = push.split()
+        # Deletes are always safe
+        if local_ref == NULL_HASH:
+            continue
+        if remote_ref == NULL_HASH:
+            remote_ref = None
+        to_push.append((remote_ref, local_ref))
+    if not to_push:
+        if VERBOSITY > QUIET:
+            print('No novel content being pushed')
+        return 0
+
+    remote_to_name, name_to_remote = remote_name_mappings()
+    if name == remote:
+        match = REMOTE_RE.match(remote)
+        if not match:
+            print_error("'{}' does not conform to a recognized target remote".format(remote))
+            return 255
+        name = remote_to_name.get('{}:{}'.format(match.group('host'), match.group('path')), None)
+        if not name:
+            print_error("Failed to map '{}' to an existing remote".format(remote))
+            return 1
+
+
+    remotes_with = {}
+    for frm, to in to_push:
+        for commit in commits_in(frm, to):
+            if commit not in remotes_with:
+                remotes_with[commit] = remotes_for(commit) or []
+            if name in remotes_with[commit]:
+                break
+    remotes_with = dict(filter(lambda pair: name not in pair[1], remotes_with.items()))
+
+    print("Verifying commits in '{}' mode".format(MODE))
+    if VERBOSITY > QUIET:
+        print('Attempting to push {} novel commits to {}...'.format(len(remotes_with), name))
+
+    need_prompt = False
+    for commit, remotes in remotes_with.items():
+        level, klass = security_level_for_commit(commit, remotes, name_to_remote=name_to_remote)
+
+        # Unkown security level of source
+        if klass == 1 and level is None and not target_level:
```

Comment:
Should we fail here if either `level` or `target_level` are not set?

```
        if klass == 1 and (level is None or not target_level):
```

Reply:
It's ok for `target_level` to be `None` (we treat a _target_ level of `None` as public). It's not OK for `level` to be `None`, though, because that indicates a commit that comes from a remote which we don't know how to categorize. Basically, pushing `None` -> `level 2` is ok, because the "most secure" `None` could possibly be (as a source) _is_ `level 2`, pushing `level 0` -> `None` is OK because the "least secure" `None` could be (as a target) is `level 0`

We're trying to avoid pushing security changes with a higher security level to remotes that have a lower security level. If the source of a commit is unknown, we must assume it is highly secretive, but if the target of a push is unknown, we must assume it is _public_